### PR TITLE
[Diagnostics]: Adds support for including external tracing information in log entries.

### LINKED ISSUE #5044: Support Istio on GKE trace spans in Google.Cloud.Diagnostics.AspNetCore logging
Hi again,

I'm attempting to associate log entries I write using `ILogger` with trace spans using logging in `Google.Cloud.Diagnostics.AspNetCore`. My environment is Istio on GKE with Istio writing traces (and generating trace IDs and span IDs).

I observe that (at least by default) Istio on GKE is using [Zipkin headers](https://github.com/openzipkin/b3-propagation) such as `X-B3-TraceId` and `X-B3-SpanId`.

Example values:

```
x-b3-traceid: 1234e1419035e455176e926b32d86285 
x-b3-spanid: 4321e68b85da75f0 
```

I note that these values are compatible with [LogEntry](https://cloud.google.com/logging/docs/reference/v2/rpc/google.logging.v2#logentry) `trace` and `span_id` (when converting `X-B3-TraceId` to a resource name).

`GoogleLogger` seems to only support the `X-Cloud-Trace-Context` header, with no ability to extend the behavior.

Ideally I would like to use ILogger and have `LogEntry`'s `trace` and `span_id` set from the Zipkin request headers. I think this would also help others using Istio on GKE.

Would it be possible to support these additional headers and set them appropriately on `trace` and `span_id` when creating `LogEntry`? Or provide an extensibility point somewhere?

Thanks!

### PR DESCRIPTION
**Pull request by amanda-tarafa**
Fixes #5044 

As always, review a commit at a time.

### CODE REVIEW

File: apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore/Logging/TraceContextForLogEntry.cs [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    {
+        /// <summary>
+        /// The trace ID to associate the log entry with.
+        /// Must not be null.
+        /// </summary>
+        public string TraceId { get; }
+
+        /// <summary>
+        /// The span ID to associate the log entry with.
+        /// Or null if there's currently no open span.
```
Comment (jskeet) [inline apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore/Logging/TraceContextForLogEntry.cs:38]:
```suggestion
        /// The span ID to associate the log entry with,
        /// or null if there's currently no open span.
```

File: apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore/Logging/TraceContextForLogEntry.cs [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+        /// </summary>
+        /// <remarks>
+        /// This is the same format that the Trace API v2 uses: a 16-character hexadecimal
```
Comment (jskeet) [inline apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore/Logging/TraceContextForLogEntry.cs:41]: Presumably that's only if that's what was passed to the constructor though - if I call `new TraceContextForLogEntry("trace-id", "this-is-not-hex")` then it's not going to complain.

File: apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore/Logging/IExternalTraceProvider.cs [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System;
+
+namespace Google.Cloud.Diagnostics.AspNetCore
+{
+    /// <summary>
+    /// Provides external trace information that can include in a log entry
```
Comment (jskeet) [inline apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore/Logging/IExternalTraceProvider.cs:20]:
```suggestion
    /// Provides external trace information to include in a log entry
```

File: apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore.IntegrationTests/Logging/LoggingTest.cs [hunk 4]
```diff
@@ -483,6 +484,78 @@ public async Task Logging_Trace_MultipleSpans()
...
+                // We have three log entries.
+                Assert.Equal(3, results.Count);
+
+                // And the resource name of the trace associated to them contains the external trace id.
+                Assert.DoesNotContain(results, entry => !entry.Trace.Contains(projectId));
+                Assert.DoesNotContain(results, entry => !entry.Trace.Contains("external_trace_id"));
+
+                // The span associated to our entry is the external span, and is the same span number
+                // as the log entry.
+                Assert.DoesNotContain(results, entry => entry.SpanId != $"external_span_number{entry.JsonPayload.Fields["message"].StringValue.Last()}");
```
Comment (jskeet) [inline apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore.IntegrationTests/Logging/LoggingTest.cs:555]: The double negative is confusing here, and despite a quick search, I still don't actually know what it means.
Is this effectively an Assert.All?

File: apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore.IntegrationTests/Logging/LoggingTest.cs [hunk 5]
```diff
@@ -961,4 +1080,11 @@ public void Invoke(Dictionary<string, string> labels)
             labels["Foo"] = "Hello, World!";
         }
     }
+
+    internal class FooExternalTraceProvider : IExternalTraceProvider
```
Comment (jskeet) [inline apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore.IntegrationTests/Logging/LoggingTest.cs:1084]: SpanCountingExternalTraceProvider?

Comment (amanda-tarafa) [reply apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore.IntegrationTests/Logging/LoggingTest.cs]: Done

File: apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore/Logging/IExternalTraceProvider.cs (hunk 3 above)
Comment (amanda-tarafa) [reply apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore/Logging/IExternalTraceProvider.cs]: Done

File: apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore/Logging/TraceContextForLogEntry.cs (hunk 1 above)
Comment (amanda-tarafa) [reply apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore/Logging/TraceContextForLogEntry.cs]: Done

File: apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore/Logging/TraceContextForLogEntry.cs (hunk 2 above)
Comment (amanda-tarafa) [reply apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore/Logging/TraceContextForLogEntry.cs]: Yes, I'm modifying the comments, as in, "it should be in this format, but we don't check"

File: apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore/Logging/TraceContextForLogEntry.cs [hunk 6]
```diff
@@ -34,12 +34,15 @@ public sealed class TraceContextForLogEntry
...
-        /// The span ID to associate the log entry with.
-        /// Or null if there's currently no open span.
+        /// The span ID to associate the log entry with,
+        /// or null if there's currently no open span.
         /// </summary>
         /// <remarks>
-        /// This is the same format that the Trace API v2 uses: a 16-character hexadecimal
-        /// encoding of an 8-byte array, such as `000000000000004a`.
+        /// For this to be matched to a span being stored in Google Cloud Trace,
+        /// it's format should be the same that the Trace API v2 uses: a 16-character hexadecimal
```
Comment (jskeet) [inline apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore/Logging/TraceContextForLogEntry.cs:42]:
```suggestion
        /// its format should be the same that the Trace API v2 uses: a 16-character hexadecimal
```

**Comment by amanda-tarafa** [comment]
Latest typo fixed. Rebased. Waiting on green before merging.

**Pull request merged**